src/main/flight/mixer.c

```diff
@@ -218,22 +218,38 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
         }
     } else {
         throttle = rcCommand[THROTTLE] - PWM_RANGE_MIN + throttleAngleCorrection;
+
 #ifdef USE_DYN_IDLE
-        if (mixerRuntime.idleMinMotorRps > 0.0f) {
-            const float maxIncrease = isAirmodeActivated() ? mixerRuntime.idleMaxIncrease : 0.04f;
-            const float minRps = rpmMinMotorFrequency();
-            const float targetRpsChangeRate = (mixerRuntime.idleMinMotorRps - minRps) * currentPidProfile->idle_adjustment_speed;
-            const float error = targetRpsChangeRate - (minRps - mixerRuntime.oldMinRps) * pidGetPidFrequency();
-            const float pidSum = constrainf(mixerRuntime.idleP * error, -currentPidProfile->idle_pid_limit, currentPidProfile->idle_pid_limit);
-            motorRangeMinIncrease = constrainf(motorRangeMinIncrease + pidSum * pidGetDT(), 0.0f, maxIncrease);
-            mixerRuntime.oldMinRps = minRps;
-            throttle += mixerRuntime.idleThrottleOffset * RC_COMMAND_THROTTLE_RANGE;
-
-            DEBUG_SET(DEBUG_DYN_IDLE, 0, motorRangeMinIncrease * 1000);
-            DEBUG_SET(DEBUG_DYN_IDLE, 1, targetRpsChangeRate);
-            DEBUG_SET(DEBUG_DYN_IDLE, 2, error);
-            DEBUG_SET(DEBUG_DYN_IDLE, 3, minRps);
-        } else {
+        if (mixerRuntime.dynIdleMinRps > 0.0f) {
+            const float maxIncrease = isAirmodeActivated() ? mixerRuntime.dynIdleMaxIncrease : 0.04f;
+            float minRps = rpmMinMotorFrequency();
+            DEBUG_SET(DEBUG_DYN_IDLE, 3, (minRps * 10));
+            const float rpsError = mixerRuntime.dynIdleMinRps - minRps;
+
+            float dynIdleP = rpsError * mixerRuntime.dynIdlePGain;
+
+            minRps = mixerRuntime.prevMinRps + 0.01f * (minRps - mixerRuntime.prevMinRps);
+            // PT1 lowpass smoothing provides smoothing and delay for D
+            // 0.01 is 6hz on 4k, 25ms delay, or 67hz at 8k with 2.5ms delay
+            // *** TODO *** correct for DTc, use proper pt1
+            float dynIdleD = (mixerRuntime.prevMinRps - minRps) * mixerRuntime.dynIdleDGain;
+
+            if (mixerRuntime.motorsIncreased) {
+                mixerRuntime.dynIdleI += (dynIdleD * 60.0f + rpsError * mixerRuntime.dynIdleIGain) * pidGetDT();
+            }
+            mixerRuntime.dynIdleI = constrainf(mixerRuntime.dynIdleI, 0.0f, maxIncrease);
+            // stop I being dragged hard negative when rps is high, hold recent value as usually closer than any other value
+
+            motorRangeMinIncrease = constrainf((dynIdleP + mixerRuntime.dynIdleI + dynIdleD), 0.0f, maxIncrease);
+            mixerRuntime.motorsIncreased = (motorRangeMinIncrease > 0);
+
+//            DEBUG_SET(DEBUG_DYN_IDLE, 0, (motorRangeMinIncrease * 10000));
+            DEBUG_SET(DEBUG_DYN_IDLE, 0, (MAX(-1000.0f, dynIdleP * 10000)));
```

Comment:
What is the reason for constraining this debug value? This just means we are wasting logging bandwidth for writing series of '1000' when we could be logging non-meaningless values instead.

Reply:
The P value, while active, is within a range of +/- 1000 of zero, approximately; that is the range we are interested in visualising in the log.  And we want good resolution to see the noise in it.
But the P value goes massively negative when rpm is above minimum, to values like -30,000, or so low as to rotate to a positive value.  Any value below -1000 is of zero interest to us, since the PIDsum will invariably be negative and the motor boost will be zero.
Without this constraint, the view in the blackbox log is of a P trace with very little vertical resolution of the band of interest, because blackbox always sets limits that include all values in the log, and if we have values at -30,000, we see little detail in the 0-500 range.  We can't fix this by changing blackbox scaling percentage either.  I have discussed this with McGyverGim.  Rather than allowing the user to configure % zoom, it would be better for the user to be allowed to set absolute minimum and maximum values for a trace.  Then this would not be necessary.

Reply:
Given that logged values are compressed, there should be less data in the log if we keep logging an identical number rather than one that changes all the time, when we don't care about the actual values.

Reply:
Good point about overruns / underrruns - pretty much all of these debug values should probably be constrained to avoid this (i.e. upper and lower bounds defined with `constrain(x, UINT16_MIN, UINT16_MAX)` or similar).

Reply:
Needs to be constrained to avoid overruns.